### PR TITLE
make `..` act like a real operator, allowing x..y ranges

### PR DESCRIPTION
**Pull request by jneen**
Also adds a `within(1..3, 2)` function to check ranges without using `CASE`.

Rebased against latest master.

~~BLOCKED behind #11.~~

### CODE REVIEW

File: lib/dentaku/ast/range.rb [hunk 1]
```diff
@@ -1,10 +1,29 @@
...
+      end
+
+      # must return something responding to ===, in this case a
+      # regular ruby Range.
+      def value
+        (left.value..right.value)
+      end
+
+      def generate_constraints!(context)
+        # pass. this is handled by AST::Case
```
Comment (joshuabates) [inline lib/dentaku/ast/range.rb]: and we don't need to do anything for when it's used as an arg in a function?

Comment (joshuabates) [reply lib/dentaku/ast/range.rb]: wait, wouldn't this mean that `field:numeric..field:text` passes the type checker?

**Review by joshuabates — APPROVED** [review]
I have some questions/comments about the lack of `Range#generate_constraints`. Looks good to merge as long as those aren't actual issues.

**Comment by jneen** [comment]
@joshuabates I've rebased this and it also includes a couple other changes, as well as actually using `generate_constraints` on `AST::Range`.

File: lib/dentaku/ast/range.rb (hunk 1 above)
Comment (jneen) [reply lib/dentaku/ast/range.rb]: you're absolutely right. i've updated this and added a few more things in following commits.

File: lib/dentaku/tokenizer.rb [hunk 2]
```diff
@@ -4,15 +4,49 @@
...
+        ']' => :close,
+        ',' => :comma,
+      }.freeze,
+      case: {
+        'case' => :open,
+        'end' => :close,
+        'then' => :then,
+        'when' => :when,
+        'else' => :else,
+        '..' => :range,
```
Comment (rtlong) [inline lib/dentaku/tokenizer.rb]: Is this meant to be under case as well as operator?

**Review by rtlong — APPROVED** [review]
Looks good in the diff against #11

File: lib/dentaku/tokenizer.rb (hunk 2 above)
Comment (jneen) [reply lib/dentaku/tokenizer.rb]: Nope. Definitely not.

**Comment by jneen** [comment]
From @joshuabates and my pairing today, we've decided we want to make `...` a parse error - currently we parse `1...2`, for example, as if it was `(1)..(.2)`, which can cause silent bugs. In fact, we have identified one fee-expression in San Diego ("building-permit-residential-multi-dwelling-unit-addition-i") that contains this bug. There were 8 projects that would have been affected but they were all guest projects, so we don't think it's necessary to retroactively fix them. We do, however, have to fix the expression, since this change would mean that expression would suddenly become invalid (unparseable).

**Pull request merged**